common/lib/capa/capa/responsetypes.py

```diff
@@ -1362,6 +1572,79 @@ def check_string_backward(self, expected, given):
             return given.lower() in [i.lower() for i in expected]
         return given in expected
 
+    def get_extended_hints(self, student_answers, new_cmap):
+        """
+        Find and install extended hints in new_cmap depending on the student answers.
+        StringResponse is probably the most complicated form we have.
+        The forms show below match in the order given, and the first matching one stops the matching.
+        <stringresponse answer="A" type="ci">
+          <correcthint>hint1</correcthint>                         <!-- hint for correct answer -->
+          <additional_answer answer="B">hint2</additional_answer>  <!-- additional_answer with its own hint -->
+          <stringequalhint answer="C">hint3</stringequalhint>      <!-- string matcher/hint for an incorrect answer -->
+          <regexphint answer="FG+">hint4</regexphint>              <!-- regex matcher/hint for an incorrect answer -->
+          <textline size="20"/>
+        </stringresponse>
+        The "ci" and "regexp" options are inherited from the parent stringresponse as appropriate.
+        """
+        if self.answer_id in student_answers:
+            student_answer = student_answers[self.answer_id]
+            # Note the atypical case of using self.id instead of self.answer_id
+            responses = self.xml.xpath('//stringresponse[@id=$id]', id=self.id)
+            if responses:
+                response = responses[0]
+
+                # First call the existing check_string to see if this is a right answer by that test.
+                # It handles the various "ci" "regexp" cases internally.
+                expected = response.get('answer').strip()
+                if self.check_string([expected], student_answer):
+                    hint_node = response.find('./correcthint')
+                    if hint_node is not None:
+                        new_cmap[self.answer_id]['msg'] += self.make_hint_div(hint_node, True)
+                    return
+
+                # Then look for additional answer with an answer= attribute
+                for node in response.findall('./additional_answer'):
+                    if self.match_hint_node(node, student_answer, self.regexp, self.case_insensitive):
+                        correct_hint = node.find('./correcthint')
+                        new_cmap[self.answer_id]['msg'] += self.make_hint_div(correct_hint, True)
+                        return
+
+                # stringequalhint and regexphint represent wrong answers
+                for node in response.findall('./stringequalhint'):
+                    if self.match_hint_node(node, student_answer, False, self.case_insensitive):
+                        new_cmap[self.answer_id]['msg'] += self.make_hint_div(node, False)
+                        return
+
+                for node in response.findall('./regexphint'):
```

Comment:
It happens once at problem-construction time, so I don't think it's too big of a problem compared to other things. The XML is kind of tiny. You can't mess with the order, since implementing a sort of hierarchy of which sort of hint has the highest priority.

Tell you what: I'll promise to instrument it on my laptop just for fun to see how much time this bit actually takes. I suspect it will be less than 1ms, but I'll get the real data. capa is not written in a high-performance way, but (1) mostly it's pretty fast anyway since these things are tiny, and (2) *this* bit is not the problem.